sae/blocks.go

```diff
@@ -5,42 +5,240 @@ package sae
 	"github.com/ava-labs/libevm/rlp"
+	"github.com/holiman/uint256"
+	"go.uber.org/zap"
 
 	"github.com/ava-labs/strevm/blocks"
+	"github.com/ava-labs/strevm/params"
+	"github.com/ava-labs/strevm/txgossip"
 )
 
-// ParseBlock parses the buffer as [rlp] encoding of a [types.Block].
-func (vm *VM) ParseBlock(context.Context, []byte) (*blocks.Block, error) {
-	ethB := new(types.Block)
-	_ = rlp.DecodeBytes(nil, ethB)
-	return nil, errUnimplemented
+func (vm *VM) newBlock(eth *types.Block, parent, lastSettled *blocks.Block) (*blocks.Block, error) {
+	return blocks.New(eth, parent, lastSettled, vm.log())
+}
+
+// ParseBlock parses the buffer as [rlp] encoding of a [types.Block]. It does
+// NOT populate the block ancestry, which is done by [VM.VerifyBlock] i.f.f.
+// verification passes.
+func (vm *VM) ParseBlock(ctx context.Context, buf []byte) (*blocks.Block, error) {
+	b := new(types.Block)
+	if err := rlp.DecodeBytes(buf, b); err != nil {
+		return nil, fmt.Errorf("rlp.DecodeBytes(..., %T): %v", b, err)
+	}
+	return vm.newBlock(b, nil, nil)
 }
 
 // BuildBlock builds a new block, using the last block passed to
-// [VM.SetPreference] as the parent and the given block context, which MAY be nil.
-func (vm *VM) BuildBlock(context.Context, *block.Context) (*blocks.Block, error) {
-	return nil, errUnimplemented
+// [VM.SetPreference] as the parent. The block context MAY be nil.
+func (vm *VM) BuildBlock(ctx context.Context, bCtx *block.Context) (*blocks.Block, error) {
+	parent := vm.preference.Load()
+
+	baseFee := uint256.NewInt(0) // TODO(arr4n) source this from the parent's worst-case *end* of execution
+	filter := txpool.PendingFilter{
+		BaseFee: baseFee,
+	}
+	return vm.buildBlock(ctx, bCtx, parent, vm.now(), vm.mempool.TransactionsByPriority(filter))
+}
+
+var errExecutionLagging = errors.New("execution lagging for settlement")
+
+// buildBlock implements the block-building logic shared by [VM.BuildBlock] and
+// [VM.VerifyBlock].
+func (vm *VM) buildBlock(ctx context.Context, bCtx *block.Context, parent *blocks.Block, blockTime time.Time, candidates []*txgossip.LazyTransaction) (*blocks.Block, error) {
+	log := vm.log().With(
+		zap.Uint64("parent_height", parent.Height()),
+		zap.Stringer("parent_hash", parent.Hash()),
+		zap.Time("block_time", blockTime),
+	)
+
+	settleAt := unix(blockTime.Add(-params.Tau))
+	lastSettled, ok, err := blocks.LastToSettleAt(settleAt, parent)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		log.Warn("Execution lagging when determining last block to settle")
+		return nil, errExecutionLagging
+	}
+
+	// Although, at the time of writing this, the implementation of
+	// [blocks.LastToSettleAt] guarantees that the returned block has finished
+	// execution, this isn't a stated invariant and it's possible for it to be
+	// more aggressive in the future.
+	wCtx, cancel := context.WithTimeout(ctx, 10*time.Millisecond)
+	defer cancel()
+	if err := lastSettled.WaitUntilExecuted(wCtx); err != nil {
```

Comment:
Even though 10ms isn't very long at all, I don't think we should be waiting here (on the consensus thread). It seems like we should instead be checking `lastSettled.Executed()` and immediately erroring or progressing.

If we are actually hitting this wait, we probably either have a bug in `WaitForEvent` or need to rethink how we select the building timestamp.